packages/shopify-app-remix/src/server/authenticate/admin/strategies/auth-code-flow.ts

```diff
@@ -36,52 +40,70 @@ export class AuthCodeFlowStrategy<
     this.logger = logger;
   }
 
-  async respondToOAuthRequests(request: Request) {
-    const {api, config} = this;
+  public async respondToOAuthRequests(request: Request): Promise<void | never> {
+    const {api, logger, config} = this;
+
     const url = new URL(request.url);
     const isAuthRequest = url.pathname === config.auth.path;
     const isAuthCallbackRequest = url.pathname === config.auth.callbackPath;
 
-    if (!isAuthRequest && !isAuthCallbackRequest) return;
+    if (isAuthRequest || isAuthCallbackRequest) {
+      const shop = api.utils.sanitizeShop(url.searchParams.get('shop')!);
+      if (!shop) throw new Response('Shop param is invalid', {status: 400});
 
-    const shop = api.utils.sanitizeShop(url.searchParams.get('shop')!);
-    if (!shop) throw new Response('Shop param is invalid', {status: 400});
+      if (isAuthRequest) {
+        throw await this.handleAuthBeginRequest(request, shop);
+      } else {
+        throw await this.handleAuthCallbackRequest(request, shop);
+      }
+    }
 
-    if (isAuthRequest) throw await this.handleAuthBeginRequest(request, shop);
+    // If this is a valid request, but it doesn't have a session token header, this is a document request. We need to
+    // ensure we're embedded if needed and we have the information needed to load the session.
+    if (!getSessionTokenHeader(request)) {
+      const params = {api, logger, config};
 
-    if (isAuthCallbackRequest) {
-      throw await this.handleAuthCallbackRequest(request, shop);
+      await this.ensureInstalledOnShop(request);
+      await ensureAppIsEmbeddedIfRequired(params, request);
```

Comment:
I don't see why either this nor `ensureSessionTokenSearchParamIfRequired` should be in the token acquisition strategy. These are things that need to happen as part of request validation, regardless of how we get the token.

It's also imo confusing for these to live in the `respondToOauthRequests` function. These have nothing to do with OAuth.

These functions should live in with the rest of the request validation logic (which is currently housed in `authenticate.ts`.